io/EptAddonWriter.cpp

```diff
@@ -0,0 +1,252 @@
+            *m_addonsArg);
+    args.add("threads", "Number of worker threads", m_numThreads);
+}
+
+void EptAddonWriter::addDimensions(PointLayoutPtr layout)
+{
+    m_nodeIdDim = layout->registerOrAssignDim("EptNodeId",
+            Dimension::Type::Unsigned32);
+    m_pointIdDim = layout->registerOrAssignDim("EptPointId",
+            Dimension::Type::Unsigned32);
+}
+
+void EptAddonWriter::ready(PointTableRef table)
+{
+    m_arbiter.reset(new arbiter::Arbiter());
+    m_pool.reset(new Pool(std::max<uint64_t>(m_numThreads, 4)));
+
+    MetadataNode meta(table.privateMetadata("ept"));
+
+    if (meta.findChild("info").value().empty())
+        throwError("EPT writer must descend from EPT reader");
+
+    const auto info(parse(meta.findChild("info").value<std::string>()));
+    const auto keys(parse(meta.findChild("keys").value<std::string>()));
+    m_hierarchyStep = meta.findChild("step").value<uint64_t>();
+    m_info.reset(new EptInfo(info));
+
+    for (const std::string s : keys.getMemberNames())
+    {
+        m_hierarchy[Key(s)] = keys[s].asUInt64();
+    }
+
+    const PointLayout& layout(*table.layout());
+    for (const std::string path : m_addonsArg->getMemberNames())
+    {
+        const auto endpoint(
+                m_arbiter->getEndpoint(arbiter::fs::expandTilde(path)));
+        const std::string dimName((*m_addonsArg)[path].asString());
+        const Dimension::Id id(layout.findDim(dimName));
+        m_addons.emplace_back(new Addon(layout, endpoint, id));
+    }
+}
+
+void EptAddonWriter::write(const PointViewPtr view)
+{
+    for (const auto& addon : m_addons)
+    {
+        log()->get(LogLevel::Debug) << "Writing addon dimension " <<
+            addon->name() << " to " << addon->ep().prefixedRoot() << std::endl;
+
+        writeOne(view, *addon);
+
+        log()->get(LogLevel::Debug) << "\tWritten" << std::endl;
+    }
+}
+
+void EptAddonWriter::writeOne(const PointViewPtr view, const Addon& addon) const
+{
+    std::vector<std::vector<char>> buffers;
+    buffers.reserve(m_hierarchy.size());
+
+    // Create an addon buffer for each node we're going to write.
+    for (const auto& p : m_hierarchy)
+    {
+        buffers.emplace_back(p.second * addon.size(), 0);
+    }
+
+    // Fill in our buffers with the data from the view.
+    PointRef pr(*view);
+    uint64_t nodeId(0);
+    uint64_t pointId(0);
+    for (uint64_t i(0); i < view->size(); ++i)
+    {
+        pr.setPointId(i);
+        nodeId = pr.getFieldAs<uint64_t>(m_nodeIdDim);
+
+        // Node IDs are 1-based to distinguish points that do not come from the
+        // EPT reader.
+        if (!nodeId) continue;
+
+        nodeId -= 1;
+        pointId = pr.getFieldAs<uint64_t>(m_pointIdDim);
+
+        auto& buffer(buffers.at(nodeId));
+        assert(pointId * addon.size() + addon.size() <= buffer.size());
+        char* dst = buffer.data() + pointId * addon.size();
+        pr.getField(dst, addon.id(), addon.type());
+    }
```

Comment:
This seems an unfortunate memory suck.  I would guess that in most instances points would be in the order read.  Can you do something like fill buffers until they hit some size and then flush them?  Can you fork a thread to flush so that you can keep processing?  Maybe that's version 2.

Reply:
I'll leave as-is for future optimization, if it becomes an issue.  Since only buffers of a single attribute are built at a time and this stage is already non-streaming, it might never be a problem.